### PR TITLE
Save upgrade test meta

### PR DESCRIPTION
**Pull request by piontec**
Adds a config option (false by default) that allows to save upgrade test results as YAML metadata.

### CODE REVIEW

File: app_test_suite/steps/scenarios/upgrade.py [hunk 1]
```diff
@@ -110,63 +115,84 @@ def pre_run(self, config: argparse.Namespace) -> None:
...
+            chart_file_name = os.path.join(d, "chart.tgz")
+            with open(chart_file_name, "wb") as f:
+                f.write(r.content)
+            TestInfoProvider().extract_chart_info(chart_file_name, CONTEXT_KEY_STABLE_CHART_YAML, context)
+        return stable_chart_ver, STABLE_APP_CATALOG_NAME, catalog_url, chart_url
 
     def run_tests(self, config: argparse.Namespace, context: Context) -> None:
         app_name = context[CONTEXT_KEY_CHART_YAML]["name"]
-        app_version = context[CONTEXT_KEY_CHART_YAML]["version"]
+        chart_version = context[CONTEXT_KEY_CHART_YAML]["version"]
```
Comment (rossf7) [inline app_test_suite/steps/scenarios/upgrade.py:162]: I'm confused by the naming here. Is this the version of the app CR or the upstream chart?

If its the version in the app CR I'd keep the current name.

File: app_test_suite/steps/scenarios/upgrade.py [hunk 2]
```diff
@@ -183,7 +209,17 @@ def run_tests(self, config: argparse.Namespace, context: Context) -> None:
...
 
-        # TODO: save upgrade metadata
+        # save metadata, if requested
+        if get_config_value_by_cmd_line_option(config, key_cfg_upgrade_save_metadata):
+            self._save_metadata(
+                app_name,
+                chart_version,
+                context[CONTEXT_KEY_CHART_YAML]["appVersion"],
+                context[CONTEXT_KEY_STABLE_CHART_YAML]["version"],
+                context[CONTEXT_KEY_STABLE_CHART_YAML]["appVersion"],
```
Comment (rossf7) [inline app_test_suite/steps/scenarios/upgrade.py:219]: Can we have a comment here explaining that we refer to appVersion as the upstream chart version?

File: app_test_suite/steps/scenarios/upgrade.py [hunk 3]
```diff
@@ -281,3 +317,32 @@ def _get_test_exec_info(self, chart_path: str, chart_ver: str, chart_config_file
...
+        chart_version: str,
+        app_version: str,
+        stable_chart_version: str,
+        stable_app_version: str,
+        cluster_type: str,
+        cluster_version: str,
+    ) -> None:
+        metadata = {
+            "appName": app_name,
+            "chartVersion": chart_version,
```
Comment (rossf7) [inline app_test_suite/steps/scenarios/upgrade.py:333]: The name is longer but I think upstreamChartVersion would be clearer here.

**Review by rossf7 — COMMENTED** [review]
Looks good. 

Just some queries on the naming of appVersion vs chartVersion.

File: app_test_suite/steps/scenarios/upgrade.py (hunk 3 above)
Comment (piontec) [reply app_test_suite/steps/scenarios/upgrade.py:333]: But it's not upstream, it's the chart's under the test.

File: app_test_suite/steps/scenarios/upgrade.py (hunk 1 above)
Comment (rossf7) [reply app_test_suite/steps/scenarios/upgrade.py:162]: Ack. I was confused why this changed but it's the chart that is under test. So aligning with that makes sense.

Same with my other comments.

**Pull request merged**